### PR TITLE
Partition worker logic into storage and compute

### PR DESCRIPTION
**Pull request by frankmcsherry**
This PR factors the logic of `dataflow::Worker` into compute and storage aspects. Each aspect is backed by the associated state, a timely worker, and the response queue. There is no other sharing that I can discern. In principle, other than a bit of ick around dataflow construction, these two entities could be dislocated and things would still work .. "great".

The design pattern is to form a subset of `Worker` for each facet, the `ActiveComputeState` and `ActiveStorageState` types, each of which wraps the subset of `Worker` they require access to. Their methods from `Worker` are instead implemented on these types.

Some rendering was extracted from `render`, which might be a mistake. Roughly, the two parts of rendering that feel the pain of having to interact (rendering the source dataflow, and rendering the rest) got pulled out and then call back in to `render` as needed. This may be an error, if we can find a good abstraction to present back at `render` and let it take control.

### Motivation

We want to prepare to actually isolate the state and logic for storage and compute, and this makes things a lot more clear (to me) without breaking anything.

The only intended functional change is that our metrics will only observe pending peeks once per scheduling quantum (time we drain the command queue) at the moment when the queue would be deepest. Previously it would also observe the pending peeks at each peek insertion, but as it only records the length it would just see a sequence of increasing numbers. Removing this meant that the `ComputeState` did not need to hold on to metric that are otherwise held by the server itself.

### Tips for Reviewers

The code movement is almost entirely the creation of the two wrapper types mentioned above, and the transportation of methods from `Worker` to them. There are many consequences mostly around `use` statements.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Comment by frankmcsherry** [comment]
Updated: The second commit moves the storage and compute dataflow building back to `render`, where its implementations are generic with respect to the new `boundary::{ComputeReplay, StorageCapture}` traits. The diff is actually smaller now because of the returned code, but the `boundary` module is new and worth inspecting (and has only a simple implementation at the moment, but could be enriched).

File: src/dataflow/src/server.rs [hunk 1]
```diff
@@ -143,6 +132,7 @@ pub fn serve(config: Config) -> Result<(Server, LocalClient), anyhow::Error> {
                 pending_peeks: Vec::new(),
                 reported_frontiers: HashMap::new(),
                 sink_metrics,
+                materialized_logger: None,
```
Comment (antiguru) [inline src/dataflow/src/server.rs:135]: Check

File: src/dataflow/src/server/compute_state.rs [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+    pub(crate) fn handle_compute_command(&mut self, cmd: ComputeCommand) {
+        match cmd {
+            ComputeCommand::CreateInstance | ComputeCommand::DropInstance => {
+                // Can be ignored for the moment.
+                // Should eventually be filtered outside of this method,
+                // as we are already deep in a specific instance.
+            }
+            ComputeCommand::CreateDataflows(_dataflows) => {
+                unreachable!("CreateDataflows should not be issued directly to compute state")
+            }
```
Comment (antiguru) [inline src/dataflow/src/server/compute_state.rs:87]: The `CreateInstance`, `DropInstance`, and `CreateDataflows` commands seem to be unicorns among the other commands as they describe orthogonal concepts. We could ponder on extracting a different `Command` variant to capture that and avoid the `unreachable!` match arms. What do you think? This is not meant to be criticism of the current PR but follow-up work.

File: src/dataflow/src/server/storage_state.rs [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+use crate::metrics::Metrics;
+use crate::render::sources::PersistedSourceManager;
+use crate::server::LocalInput;
+use crate::source::metrics::SourceBaseMetrics;
+use crate::source::timestamp::TimestampBindingRc;
+use crate::source::SourceToken;
+
+/// How frequently each dataflow worker sends timestamp binding updates
+/// back to the coordinator.
+static TS_BINDING_FEEDBACK_INTERVAL_MS: u128 = 1_000;
```
Comment (antiguru) [inline src/dataflow/src/server/storage_state.rs:35]: `Duration` constructors are `const` since Rust 1.58, so we could change the type to `const Duration` here. Can do in a separate PR.

**Review by antiguru — APPROVED** [review]
Thanks, this looks much more organized.

File: src/dataflow/src/server.rs (hunk 1 above)
Comment (antiguru) [reply src/dataflow/src/server.rs:135]: Is OK.

File: src/dataflow/src/server/compute_state.rs (hunk 2 above)
Comment (frankmcsherry) [reply src/dataflow/src/server/compute_state.rs:87]: Ah, I'm not sure yet. They are each new enough that idk what they end up looking like. Probably the create and drop instance commands don't survive here, but .. create dataflows would probably be fine except that we need to pick out the source creation from the dataflow, a thing that could/should actually happen with `StorageCommand::CreateSource`, or by compute pinging storage on construction.

File: src/dataflow/src/server/storage_state.rs (hunk 3 above)
Comment (frankmcsherry) [reply src/dataflow/src/server/storage_state.rs:35]: Cool. I just copy/pasted but happy to adapt / have someone else adapt.

**Comment by frankmcsherry** [comment]
Thank you very much for the review!
